### PR TITLE
Migrate ws2_32 APIs

### PR DESCRIPTION
**Pull request by skdltmxn**

### CODE REVIEW

**Comment by skdltmxn** [comment]
Now I think it's safer :)

**Comment by skdltmxn** [comment]
Fixed.

File: src/shared/mstcpip.rs [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+    RCVALL_SOCKETLEVELONLY = 2,
+    RCVALL_IPLEVEL = 3,
+}}
+pub type PRCVALL_VALUE = *mut RCVALL_VALUE;
+STRUCT!{struct RCVALL_IF {
+    Mode: RCVALL_VALUE,
+    Interface: ULONG,
+}}
+pub type PRCVALL_IF = *mut RCVALL_IF;
+pub const TCP_INITIAL_RTO_UNSPECIFIED_RTT: USHORT = -1i32 as USHORT;
```
Comment (retep998) [inline src/shared/mstcpip.rs]: `-1i16 as u16`. See my latest changes to `CONTRIBUTING.md`.

File: src/shared/ws2def.rs [hunk 2]
```diff
@@ -5,14 +5,14 @@
 // All files in the project carrying such notice may not be copied, modified, or distributed
 // except according to those terms.
 //! Core definitions for the Winsock2 specification
-
-use ctypes::{__int64, c_char, c_int, c_short, c_void};
+#![allow(overflowing_literals)]
```
Comment (retep998) [inline src/shared/ws2def.rs]: This can be removed.

File: src/um/winsock2.rs [hunk 3]
```diff
@@ -165,30 +226,181 @@ STRUCT!{struct linger {
...
+pub const FD_QOS_BIT: c_long = 6;
+pub const FD_QOS: c_long = 1 << FD_QOS_BIT;
+pub const FD_GROUP_QOS_BIT: c_long = 7;
+pub const FD_GROUP_QOS: c_long = 1 << FD_GROUP_QOS_BIT;
+pub const FD_ROUTING_INTERFACE_CHANGE_BIT: c_long = 8;
+pub const FD_ROUTING_INTERFACE_CHANGE: c_long = 1 << FD_ROUTING_INTERFACE_CHANGE_BIT;
+pub const FD_ADDRESS_LIST_CHANGE_BIT: c_long = 9;
+pub const FD_ADDRESS_LIST_CHANGE: c_long = 1 << FD_ADDRESS_LIST_CHANGE_BIT;
+pub const FD_MAX_EVENTS: usize = 10;
+pub const FD_ALL_EVENTS: c_long = 1 << (FD_MAX_EVENTS as c_long) - 1;
```
Comment (retep998) [inline src/um/winsock2.rs]: Don't forget this cast.

File: src/um/winsock2.rs [hunk 4]
```diff
@@ -165,30 +226,181 @@ STRUCT!{struct linger {
...
 pub type LPWSAEVENT = LPHANDLE;
 pub type WSAOVERLAPPED = OVERLAPPED;
 pub type LPWSAOVERLAPPED = *mut OVERLAPPED;
-pub const WSA_IO_PENDING: DWORD = ERROR_IO_PENDING;
-pub const WSA_IO_INCOMPLETE: DWORD = ERROR_IO_INCOMPLETE;
-pub const WSA_INVALID_HANDLE: DWORD = ERROR_INVALID_HANDLE;
-pub const WSA_INVALID_PARAMETER: DWORD = ERROR_INVALID_PARAMETER;
-pub const WSA_NOT_ENOUGH_MEMORY: DWORD = ERROR_NOT_ENOUGH_MEMORY;
-pub const WSA_OPERATION_ABORTED: DWORD = ERROR_OPERATION_ABORTED;
+pub const WSA_IO_PENDING: c_int = ERROR_IO_PENDING as c_int;
```
Comment (retep998) [inline src/um/winsock2.rs]: And these.

File: src/um/ws2tcpip.rs [hunk 5]
```diff
@@ -3,16 +3,345 @@
...
+    LPWSAOVERLAPPED, LPWSAOVERLAPPED_COMPLETION_ROUTINE, SOCKET, WSA_NOT_ENOUGH_MEMORY, timeval,
+};
+use vc::vcruntime::size_t;
+pub const UDP_NOCHECKSUM: c_int = 1;
+pub const UDP_CHECKSUM_COVERAGE: c_int = 20;
+pub const EAI_AGAIN: DWORD = WSATRY_AGAIN;
+pub const EAI_BADFLAGS: DWORD = WSAEINVAL;
+pub const EAI_FAIL: DWORD = WSANO_RECOVERY;
+pub const EAI_FAMILY: DWORD = WSAEAFNOSUPPORT;
+pub const EAI_MEMORY: DWORD = WSA_NOT_ENOUGH_MEMORY as DWORD;
```
Comment (retep998) [inline src/um/ws2tcpip.rs]: And this.

Comment (skdltmxn) [inline src/um/ws2tcpip.rs]: Oh, there were more, okay I will fix those.
By the way, I can't really find which part of the new CONTRIBUTING.md is about the casting.
Can you point it out please?

Comment (retep998) [inline src/um/ws2tcpip.rs]: https://github.com/retep998/winapi-rs/blob/dev/CONTRIBUTING.md#constants

I accidentally forgot to commit and push those changes until just now...

File: src/um/winsock2.rs [hunk 6]
```diff
@@ -165,30 +226,181 @@ STRUCT!{struct linger {
...
-pub const WSA_INVALID_PARAMETER: DWORD = ERROR_INVALID_PARAMETER;
-pub const WSA_NOT_ENOUGH_MEMORY: DWORD = ERROR_NOT_ENOUGH_MEMORY;
-pub const WSA_OPERATION_ABORTED: DWORD = ERROR_OPERATION_ABORTED;
+pub const WSA_IO_PENDING: c_int = ERROR_IO_PENDING as i32;
+pub const WSA_IO_INCOMPLETE: c_int = ERROR_IO_INCOMPLETE as i32;
+pub const WSA_INVALID_HANDLE: c_int = ERROR_INVALID_HANDLE as i32;
+pub const WSA_INVALID_PARAMETER: c_int = ERROR_INVALID_PARAMETER as i32;
+pub const WSA_NOT_ENOUGH_MEMORY: c_int = ERROR_NOT_ENOUGH_MEMORY as i32;
+pub const WSA_OPERATION_ABORTED: c_int = ERROR_OPERATION_ABORTED as i32;
+pub const WSA_INVALID_EVENT: WSAEVENT = 0 as WSAEVENT;
```
Comment (skdltmxn) [inline src/um/winsock2.rs]: In this case, WSA_INVALID_EVENT is a pointer but it's not negative.
Should I still cast it using WSAEVENT?

Comment (skdltmxn) [inline src/um/winsock2.rs]: In this case, WSA_INVALID_EVENT is a pointer but it's not negative.
Should I still cast it using WSAEVENT?

Comment (retep998) [inline src/um/winsock2.rs]: Yes.

File: src/um/winsock2.rs [hunk 7]
```diff
@@ -165,30 +226,181 @@ STRUCT!{struct linger {
...
+pub const FD_QOS_BIT: c_long = 6;
+pub const FD_QOS: c_long = 1 << FD_QOS_BIT;
+pub const FD_GROUP_QOS_BIT: c_long = 7;
+pub const FD_GROUP_QOS: c_long = 1 << FD_GROUP_QOS_BIT;
+pub const FD_ROUTING_INTERFACE_CHANGE_BIT: c_long = 8;
+pub const FD_ROUTING_INTERFACE_CHANGE: c_long = 1 << FD_ROUTING_INTERFACE_CHANGE_BIT;
+pub const FD_ADDRESS_LIST_CHANGE_BIT: c_long = 9;
+pub const FD_ADDRESS_LIST_CHANGE: c_long = 1 << FD_ADDRESS_LIST_CHANGE_BIT;
+pub const FD_MAX_EVENTS: usize = 10;
+pub const FD_ALL_EVENTS: c_long = 1 << (FD_MAX_EVENTS as i32) - 1;
```
Comment (retep998) [inline src/um/winsock2.rs]: Actually, is this cast even necessary? I see `Shl` implementations for *every* combination of two integer types.

Comment (skdltmxn) [inline src/um/winsock2.rs]: Oh, there was even precedence error here, fixed!

File: src/um/winsock2.rs (hunk 6 above)
Comment (skdltmxn) [inline src/um/winsock2.rs]: WIthout casting I get compile error.

File: src/um/winsock2.rs [hunk 8]
```diff
@@ -5,27 +5,37 @@
 // All files in the project carrying such notice may not be copied, modified, or distributed
 // except according to those terms.
 //! Definitions to be used with the WinSock 2 DLL and WinSock 2 applications.
-
-use ctypes::{ __uint64, c_char, c_int, c_long, c_short, c_uchar, c_uint, c_ulong, c_ushort};
-use shared::basetsd::{UINT_PTR, ULONG_PTR};
+#![allow(overflowing_literals)]
```
Comment (retep998) [inline src/um/winsock2.rs]: This can be removed.

**Comment by skdltmxn** [comment]
Removed.

**Comment by Susurrus** [comment]
🎉

**Pull request merged**